pkg/decoder/message_producer.go

```diff
@@ -0,0 +1,178 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2017 Datadog, Inc.
+
+package decoder
+
+import (
+	"bytes"
+	"errors"
+	"regexp"
+	"sync"
+	"time"
+)
+
+// messageProducer consumes new lines to compute and emit messages
+// Prepare and Dispose can be used for special needs
+type messageProducer interface {
```

Comment:
I really don't feel like this is behaving as a producer, but rather as a multiline buffer / delimiter. Perhaps the naming could be improved upon ?